chain/actor.go

```diff
@@ -0,0 +1,240 @@
+		} else if p.lastTipSet.Height() < ts.Height() {
+			// last tipset seen was the parent
+			data, report, err = p.processStateChanges(ctx, ts, p.lastTipSet, stateTree, p.lastStateTree)
+		} else {
+			log.Errorw("out of order tipsets", "height", ts.Height(), "last_height", p.lastTipSet.Height())
+		}
+	}
+
+	p.lastTipSet = ts
+	p.lastStateTree = stateTree
+
+	if err != nil {
+		log.Errorw("error received while processing actors, closing lens", "error", err)
+		if cerr := p.Close(); cerr != nil {
+			log.Errorw("error received while closing lens", "error", cerr)
+		}
+	}
+	return data, report, err
+}
+
+func (p *ActorStateProcessor) processStateChanges(ctx context.Context, ts *types.TipSet, pts *types.TipSet, stateTree *state.StateTree, parentStateTree *state.StateTree) (model.PersistableWithTx, *visormodel.ProcessingReport, error) {
+	log.Debugw("processing state changes", "height", ts.Height(), "parent_height", pts.Height())
+
+	report := &visormodel.ProcessingReport{
+		Height:    int64(ts.Height()),
+		StateRoot: ts.ParentState().String(),
+		Status:    visormodel.ProcessingStatusOK,
+	}
+
+	if !types.CidArrsEqual(ts.Parents().Cids(), pts.Cids()) {
+		report.ErrorsDetected = xerrors.Errorf("child is not on the same chain")
+		return nil, report, nil
+	}
+
+	changes, err := state.Diff(parentStateTree, stateTree)
+	if err != nil {
+		report.ErrorsDetected = xerrors.Errorf("failed to diff state trees: %w", err)
+		return nil, report, nil
+	}
+
+	ll := log.With("height", int64(ts.Height()))
+
+	ll.Debugw("found actor state changes", "count", len(changes))
+
+	start := time.Now()
+
+	// Run each task concurrently
+	results := make(chan *ActorStateResult, len(changes))
+	for addr, act := range changes {
+		go p.runActorStateExtraction(ctx, ts, pts, addr, act, results)
+	}
+
+	data := make(PersistableWithTxList, 0, len(changes))
+	errorsDetected := make([]*ActorStateError, 0, len(changes))
+	skippedActors := 0
+
+	// Gather results
+	inFlight := len(changes)
+	for inFlight > 0 {
+		res := <-results
+		inFlight--
+		elapsed := time.Since(start)
+		lla := log.With("height", int64(ts.Height()), "actor", actorstate.ActorNameByCode(res.Code), "address", res.Address)
+
+		if res.Error != nil {
+			lla.Errorw("actor returned with error", "error", res.Error.Error())
+			errorsDetected = append(errorsDetected, &ActorStateError{
+				Code:    res.Code.String(),
+				Name:    actorstate.ActorNameByCode(res.Code),
+				Head:    res.Head.String(),
+				Address: res.Address,
+				Error:   res.Error.Error(),
+			})
+			continue
+		}
+
+		if res.SkippedParse {
+			lla.Debugw("skipped actor without extracter")
+			skippedActors++
+		}
+
+		lla.Debugw("actor returned with data", "time", elapsed)
+		data = append(data, res.Data)
+	}
+
+	if skippedActors > 0 {
+		report.StatusInformation = fmt.Sprintf("did not parse %d actors", skippedActors)
+	}
+
+	if len(errorsDetected) != 0 {
+		report.ErrorsDetected = errorsDetected
+	}
```

Comment:
can you append errors on the report directly, rather than in a separate variable you then append to the report here?

Reply:
Done